autorest/azure/async_test.go

```diff
@@ -0,0 +1,516 @@
+	if !(OperationResource{Status: OperationCanceled}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a canceled operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsTrueIfFailed(t *testing.T) {
+	if !(OperationResource{Status: OperationFailed}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a failed operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsTrueIfSuccessful(t *testing.T) {
+	if !(OperationResource{Status: OperationSucceeded}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a successful operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsFalseNotTerminated(t *testing.T) {
+	if (OperationResource{Status: "UnknownStatus"}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated returned true for a non-terminal operation")
+	}
+}
+
+func TestOperationResource_GetErrorReturnsErrorIfCanceled(t *testing.T) {
+	if (OperationResource{Status: OperationCanceled}).GetError() == nil {
+		t.Errorf("azure: OperationResource#GetError failed to return an error for canceled operations")
+	}
+}
+
+func TestOperationResource_GetErrorReturnsErrorIfFailed(t *testing.T) {
+	if (OperationResource{Status: OperationFailed}).GetError() == nil {
+		t.Errorf("azure: OperationResource#GetError failed to return an error for failed operations")
+	}
+}
+
+func TestOperationResource_GetErrorDoesReturnUnlessFailedOrCanceled(t *testing.T) {
+	if (OperationResource{Status: OperationSucceeded}).GetError() != nil {
+		t.Errorf("azure: OperationResource#GetError return an error for operation that was not canceled or failed")
+	}
+}
+
+func TestDoPollForAsynchronous_IgnoresUnspecifiedStatusCodes(t *testing.T) {
+	client := mocks.NewSender()
+
+	r, _ := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Duration(0), time.Duration(0)))
+
+	if client.Attempts() != 1 {
+		t.Errorf("azure: DoPollForAsynchronous polled for unspecified status code")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_PollsForSpecifiedStatusCodes(t *testing.T) {
+	client := mocks.NewSender()
+	client.AppendResponse(newAsynchronousResponse())
+
+	r, _ := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond, time.Millisecond))
+
+	if client.Attempts() != 2 {
+		t.Errorf("azure: DoPollForAsynchronous failed to poll for specified status code")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_CanBeCanceled(t *testing.T) {
+	cancel := make(chan struct{})
+	delay := 5 * time.Second
+
+	client := mocks.NewSender()
+	client.AppendResponse(newAsynchronousResponse())
+	client.AppendAndRepeatResponse(newOperationResourceResponse("Busy"), -1)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	start := time.Now()
+	go func() {
+		wg.Done()
```

Comment:
@ahmetalpbalkan That approach, in this case, requires internal knowledge of how DoPollForAsynchronous works. This approach will work in all cases.